org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -495,25 +496,47 @@ public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
 				if (currentLS != null && currentLS == LanguageServerWrapper.this.languageServer) {
 					currentLS.getWorkspaceService().didChangeWorkspaceFolders(new DidChangeWorkspaceFoldersParams(wsFolderEvent));
 				}
-				ResourcesPlugin.getWorkspace().addResourceChangeListener(workspaceFolderUpdater, IResourceChangeEvent.POST_CHANGE);
+				ResourcesPlugin.getWorkspace().addResourceChangeListener(workspaceFolderUpdater, IResourceChangeEvent.POST_CHANGE | IResourceChangeEvent.PRE_DELETE);
 				return Status.OK_STATUS;
 			}
 		}.schedule();
 	}
 
-	private static final @Nullable WorkspaceFoldersChangeEvent toWorkspaceFolderEvent(IResourceChangeEvent e) {
-		if (e.getType() != IResourceChangeEvent.POST_CHANGE) {
+	private static final @Nullable WorkspaceFoldersChangeEvent toWorkspaceFolderEvent(
+			IResourceChangeEvent e) {
+
+		// If a project delete then the delta is null, but we get the project in the top-level resource
+		WorkspaceFoldersChangeEvent wsFolderEvent = new WorkspaceFoldersChangeEvent();
+		if (e.getType() == IResourceChangeEvent.PRE_DELETE) {
+			final IResource resource = e.getResource();
+			if (resource instanceof IProject) {
+				wsFolderEvent.getRemoved()
+						.add(LSPEclipseUtils.toWorkspaceFolder((IProject)resource));
+				return wsFolderEvent;
+			} else {
+				return null;
+			}
+		} else if (e.getType() != IResourceChangeEvent.POST_CHANGE) {
```

Comment:
could we do this at the very beginning of the method, so that if this condition is met, we do not need to create an WorkspaceFoldersChangeEvent that we will never use?

Reply:
I am also a relative newcomer to programming on the Eclipse platform. Normally I would code as by API and remove the guard, since it should never be hit. According to git history @mickaelistria  introduced the guard, so maybe he has some insight why he added it in the first place.

Reply:
> if the subscription at line 499 is honoured by Eclipse it should not to be needed, is that correct?

Correct. However, it's always safer to re-validate important assumptions in the code. Adding a check has no significant impact on the performance but has a very positive impact on our capacity to change the code safely, so overall it's beneficial.

Reply:
Thanks. Noted for future!